src/codegen/compile.ml

```diff
@@ -10352,97 +10766,116 @@ and compile_prim_invocation (env : E.t) ae p es at =
 
     | (Nat8|Nat16), Nat ->
       SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 t1) e ^^
       Prim.prim_shiftWordNtoUnsigned env (TaggedSmallWord.shift_of_type t1)
 
     | (Int8|Int16), Int ->
       SR.Vanilla,
-      compile_exp_vanilla env ae e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 t1) e ^^
       Prim.prim_shiftWordNtoSigned env (TaggedSmallWord.shift_of_type t1)
 
     | Nat32, Nat ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 Nat32) e ^^
       Prim.prim_word32toNat env
 
     | Int32, Int ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
+      compile_exp_as env ae (SR.UnboxedWord32 Int32) e ^^
       Prim.prim_word32toInt env
 
     | Nat64, Nat ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord64 e ^^
+      compile_exp_as env ae (SR.UnboxedWord64 Nat64) e ^^
       BigNum.from_word64 env
 
     | Int64, Int ->
       SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord64 e ^^
+      compile_exp_as env ae (SR.UnboxedWord64 Int64) e ^^
       BigNum.from_signed_word64 env
 
     | Nat32, Char ->
-      SR.Vanilla,
-      compile_exp_as env ae SR.UnboxedWord32 e ^^
-      TaggedSmallWord.check_and_tag_codepoint env
+      SR.UnboxedWord32 Type.Char, (* ! *)
+      compile_exp_as env ae (SR.UnboxedWord32 Nat32) e ^^
+      TaggedSmallWord.check_and_msb_adjust_codepoint env (* TBR *)
 
     | Float, Int ->
       SR.Vanilla,
       compile_exp_as env ae SR.UnboxedFloat64 e ^^
-      E.call_import env "rts" "bigint_of_float64"
+      E.call_import env "rts" "bigint_of_float64" ^^
+      let set_b, get_b = new_local env "b" in
+      (* adjust 31 to ubit scalar, done here to avoid touching bigint.rs *)
+      set_b ^^
+      get_b ^^
+      BitTagged.if_tagged_scalar env [I32Type]
+        (get_b ^^
+         compile_shrS_const 1l ^^
+         BigNum.from_signed_word32 env)
+        (get_b)
```

Comment:
The analogous logic is used in `compile_load_from_data_buf` (covering both signed and unsigned cases). Maybe, this logic could be extracted as a common "bigint result sanitizer" helper function.

Reply:
I've removed the need for this logic entirely by getting the rts to call a new compile.ml exported function for constructing a motoko `Int` from an i32 (`int_from_i32`), boxing only when necessary.
That actually revealed a subtle bug in `compile_load_from_data_buf` which was mistakenly calling `Num.from_(signed)_word32`, not `BigNum.from_(signed)_word32`, causing it to heap allocate even small deserialized bignums, undoing @ggreif's hard work to optimize this.

The change fixes the regression in idl.mo.